File: C(++)/[2 sem] 2DWorldSimulator/code/src/worldObjects/animals/Turtle.cpp
//
// Created by szymo on 4/18/2023.
//

#include ".headers/Turtle.h"
#include "../../utils/helpFuncs.h"
#include "iostream"

#define TURTLE_STRENGTH 2
#define TURTLE_INITIATIVE 1
#define TURTLE_SYMBOL 'T'
#define STAY_CHANCE 75

Turtle::Turtle(int x, int y, World *world) : Animal(x, y, world, TURTLE_STRENGTH, TURTLE_INITIATIVE, TURTLE_SYMBOL){}

Turtle::~Turtle() = default;

void Turtle::action() {
    int stay = random(0,100);
    if (stay > STAY_CHANCE) { Animal::action(); }
}

Organism* Turtle::collision(Organism *attacker) {
    if (attacker->getType() == "Animal" && this->getSymbol() == attacker->getSymbol())
    {
        this->multiply(this->x, this->y);
        return nullptr;
    }
    if (attacker->getStrength() >= 5) {return attacker;}
    else {
        // attacker gets back to previous position, so there's no need to return + move it
        return nullptr;
    }
}

std::string Turtle::getSpecieName() const {
    return "Turtle";
}

File: Java/[2 sem] 2DWorldSimulator/code/Turtle.java
import java.util.Objects;
import java.util.Random;

public class Turtle extends Animal {
    private static final int TURTLE_STRENGTH = 2;
    private static final int TURTLE_INITIATIVE = 1;
    private static final char TURTLE_SYMBOL = 'T';
    private static final int STAY_CHANCE = 75;

    public Turtle(int x, int y, World world) {
        super(x, y, world, TURTLE_STRENGTH, TURTLE_INITIATIVE, TURTLE_SYMBOL, 0);
    }

    @Override
    public void action() {
        int stay = new Random().nextInt(101);
        if (stay > STAY_CHANCE) { super.action(); }
    }

    @Override
    public Organism collision(Organism attacker) {
        if (attacker.getType().equals("Animal") && Objects.equals(this.getSymbol(), attacker.getSymbol())) {
            this.multiply(this.getX(), this.getY());
            return new Turtle(0, 0, this.world);
        }
        if (attacker.getStrength() >= 5) {
            return attacker;
        } else {
            // attacker gets back to the previous position, so there's no need to return + move it
            return null;
        }
    }

    @Override
    public String getSpecieName() {
        return "Turtle";
    }
}